### PR TITLE
opencv4: cuda: fix libstdc++ errors downstream, plus upkeep

### LINKED ISSUE #258327: `opencv4` and `scipy` cause `glibc` version conflicts with `cudaSupport`
### Describe the bug

Trying to import `opencv4` and `scipy` with `cuda` can cause `glibc` version conflicts.

### Steps To Reproduce

```nix
{
  inputs.nixpkgs.url = "github:nixos/nixpkgs/0396d3b0fb7f62ddc79a506ad3e6124f01d2ed0a";

  outputs = { nixpkgs, ... }: let
    reproduction = { pkgs, ... }: {
      nixpkgs.config = {
        allowUnfree = true;
        cudaSupport = true;

        # the cudaCapabilities line is probably not required for reproduction,
        # but I specify just a single capability version to reduce build time
        cudaCapabilities = [ "8.6" ];
      };

      environment.systemPackages = [(pkgs.python3.withPackages (ps: [
        ps.opencv4
        ps.scipy
      ]))];
    };
  in
  {
    nixosConfigurations.testing = nixpkgs.lib.nixosSystem {
      system = "x86_64-linux";
      modules = [
        { users.users.root.password = ""; }
        reproduction
      ];
    };
  };
}
```

```sh
nix build .#nixosConfigurations.testing.config.system.build.vm
result/bin/run-nixos-vm -nographic
```

Login as `root` (the password is empty) and then run

```sh
python -c "import cv2" # this is OK
python -c "import scipy.fft" # this is OK
python -c "import scipy.fft; import cv2" # this is also OK

python -c "import cv2; import scipy.fft" # Boom!
```

```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/nix/store/5phig9ydshwkaka96xn6llxpfcsjyzrl-python3-3.10.12-env/lib/python3.10/site-packages/scipy/fft/__init__.py", line 92, in <module>
    from ._helper import next_fast_len
  File "/nix/store/5phig9ydshwkaka96xn6llxpfcsjyzrl-python3-3.10.12-env/lib/python3.10/site-packages/scipy/fft/_helper.py", line 3, in <module>
    from ._pocketfft import helper as _helper
  File "/nix/store/5phig9ydshwkaka96xn6llxpfcsjyzrl-python3-3.10.12-env/lib/python3.10/site-packages/scipy/fft/_pocketfft/__init__.py", line 3, in <module>
    from .basic import *
  File "/nix/store/5phig9ydshwkaka96xn6llxpfcsjyzrl-python3-3.10.12-env/lib/python3.10/site-packages/scipy/fft/_pocketfft/basic.py", line 6, in <module>
    from . import pypocketfft as pfft
ImportError: /nix/store/pfx4gg15nllsa6cwfhjnink1jr35dpfq-gcc-11.3.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /nix/store/5phig9ydshwkaka96xn6llxpfcsjyzrl-python3-3.10.12-env/lib/python3.10/site-package)
```

### Expected behavior

No crashes

### Notify maintainers

I am not really sure, which of the 3 components in the reproduction is actually responsible (if any), so I am pinging the maintainers of all three.

cuda: @connorbaker @samuela @SomeoneSerge @NixOS/cuda-maintainers
opencv4: @basvandijk
scipy: @fridh @doronbehar

Also, this might be related to #220341 #226165

### PR DESCRIPTION
**Pull request by SomeoneSerge**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/258327

Adds a test for and fixes the following libstdc++ mismatch error:

```
python3.11-pytlsd> ==================================== ERRORS ====================================
python3.11-pytlsd> _______________________ ERROR collecting tests/tests.py ________________________
python3.11-pytlsd> ImportError while importing test module '/build/source/tests/tests.py'.
python3.11-pytlsd> Hint: make sure your test modules/packages have valid Python names.
python3.11-pytlsd> Traceback:
python3.11-pytlsd> /nix/store/ffll6glz3gwx342z0ch8wx30p5cnqz1z-python3-3.11.5/lib/python3.11/importlib/__init__.py:126: in import_module
python3.11-pytlsd>     return _bootstrap._gcd_import(name[level:], package, level)
python3.11-pytlsd> tests/tests.py:7: in <module>
python3.11-pytlsd>     from skimage.transform import pyramid_reduce
python3.11-pytlsd> /nix/store/zr2pijvixax0say6n6d1s0mg3f970xv2-python3.11-scikit-image-0.21.0/lib/python3.11/site-packages/skimage/transform/__init__.py:38: in <module>
python3.11-pytlsd>     from .radon_transform import (radon, iradon, iradon_sart,
python3.11-pytlsd> /nix/store/zr2pijvixax0say6n6d1s0mg3f970xv2-python3.11-scikit-image-0.21.0/lib/python3.11/site-packages/skimage/transform/radon_transform.py:3: in <module>
python3.11-pytlsd>     from scipy.interpolate import interp1d
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/interpolate/__init__.py:167: in <module>
python3.11-pytlsd>     from ._interpolate import *
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/interpolate/_interpolate.py:12: in <module>
python3.11-pytlsd>     from . import _fitpack_py
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/interpolate/_fitpack_py.py:10: in <module>
python3.11-pytlsd>     from ._bsplines import BSpline
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/interpolate/_bsplines.py:9: in <module>
python3.11-pytlsd>     from scipy.optimize import minimize_scalar
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/optimize/__init__.py:422: in <module>
python3.11-pytlsd>     from ._linprog import linprog, linprog_verbose_callback
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/optimize/_linprog.py:21: in <module>
python3.11-pytlsd>     from ._linprog_highs import _linprog_highs
python3.11-pytlsd> /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/optimize/_linprog_highs.py:20: in <module>
python3.11-pytlsd>     from ._highs._highs_wrapper import _highs_wrapper
python3.11-pytlsd> E   ImportError: /nix/store/n0gf2wl3g6229q275r4cxs2yhnam0xvg-gcc-11.4.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /nix/store/nab40hl5f2hk4i1x5rlkimmn20dyqh40-python3.11-scipy-1.11.3/lib/python3.11/site-packages/scipy/optimize/_highs/_highs_wrapper.cpython-311-x86_64-linux-gnu.so)
python3.11-pytlsd> =========================== short test summary info ============================
python3.11-pytlsd> ERROR tests/tests.py
python3.11-pytlsd> !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
python3.11-pytlsd> =============================== 1 error in 0.28s ===============================
python3.11-pytlsd> /nix/store/bbxdw4rgwwl3gnajri82yidr1nlsfskf-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/jfpbrdgvfr2qbac02lml42p1kvh4p1nf-python3.11-pytlsd-0.0.5.drv' failed with exit code 2;
```

...observed when importing `cv2` (built with cuda support and the older nvcc-compatible gcc) before importing a native python extension from nixpkgs (which uses the newer libstdc++) like so: https://github.com/iago-suarez/pytlsd/blob/81e0fee35f1fdaf546ec73c71ad3ed488c8ee83b/tests/tests.py#L3-L7

The test passes for the opencv3 attribute because it's just built with the newer gcc(12). I decided not to touch it further, but my expectation is that `opencv3` with CUDA will start failing once we bump gcc.

The tests can be run as follows:

```
❯ nix build -f . --arg config '{ cudaSupport = true; cudaCapabilities = [ "8.6" ]; enableForwardCompat = false; allowUnfree = true; }' opencv4.tests.no-libstdcxx-errors
❯ nix build -f . --arg config '{ cudaSupport = true; cudaCapabilities = [ "8.6" ]; enableForwardCompat = false; allowUnfree = true; }' opencv3.tests.no-libstdcxx-errors
```

You can check out the first commit and verify that `opencv4.no-libstdcxx-errors` fails until it's updated in the next commit

In the longer run we should kill `backendStdenv` and instead implement https://github.com/NixOS/nixpkgs/issues/226165

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

@NixOS/cuda-maintainers

### CODE REVIEW

File: pkgs/development/libraries/opencv/4.x.nix [hunk 1]
```diff
@@ -37,7 +37,7 @@
 , blas
 , enableContrib ? true
 
-, enableCuda ? config.cudaSupport && stdenv.hostPlatform.isx86_64
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/opencv/4.x.nix:40]: Oh I dropped this without thinking. Why do we ignore `config` on non-x86_64 in the first place?

Comment (SomeoneSerge) [reply pkgs/development/libraries/opencv/4.x.nix:40]: https://github.com/NixOS/nixpkgs/commit/c4adeddb5f8e945517068968d06ea838b7c24bd3

> config.cudaSupport can be true and cudatoolkit doesn't work with i686,
> which can happen inside pkgsi686Linux wrapping.

OK, so this wasn't about aarch64 😅

Comment (SomeoneSerge) [reply pkgs/development/libraries/opencv/4.x.nix:40]: Also, why would anyone set `config.cudaSupport = true` on an i686 device?

Comment (SomeoneSerge) [reply pkgs/development/libraries/opencv/4.x.nix:40]: I should move this to a separate commit though

**Review by samuela — APPROVED** [review]
diff LGTM!

**Pull request merged**